File: nprimo/go/versao2/main.go
package main

import (
	"fmt"
	"time"
)

func main() {
	start := time.Now()
	primes := make([]int, 0)
	for i := 2; i <= 1000000; i++ {
		if i%2 == 0 {
			if i == 2 {
				 primes = append(primes, i)
			} else {
				continue
			}
		}

		isPrime := true
		for j := 3; j*j <= i; j += 2 {
			if i%j == 0 {
				isPrime = false
				break
			}
		}

		if isPrime {
			primes = append(primes, i)
			fmt.Println(i)
		}
	}

	for _, prime := range primes {
		fmt.Println(prime)
	}

	end := time.Now()
	finish := end.Sub(start)
	fmt.Println("time:", finish)
}

File: nprimo/rust/versao2/main.rs
use std::time::Instant;

fn sieve_of_eratosthenes(n: usize) -> Vec<usize> {
    let mut primes = vec![true; n + 1];
    let mut result = Vec::new();

    primes[0] = false;
    primes[1] = false;

    for p in 2..=n {
        if primes[p] {
            result.push(p);
            for i in (p * p..=n).step_by(p) {
                primes[i] = false;
            }
        }
    }

    result
}

fn main() {
    let n = 1000000;
    let start = Instant::now();
    let primes = sieve_of_eratosthenes(n);
    let duration = start.elapsed();

    println!("Prime numbers up to {} are:", n);
    for prime in &primes {
        println!("{}", prime);
    }

    // println!("Prime numbers up to {} are: {:?}", n, primes);
    println!("Time: {:?}", duration);
}

